pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,831 @@
+			"PSP22::decrease_allowance",
+			vec![spender.to_string(), value.to_string()],
+			None,
+		)
+	}
+
+	/// PSP22Metadata test utilities.
+
+	pub(super) fn token_name(session: &mut Session<Sandbox>) -> Option<String> {
+		call::<Option<String>>(session, "PSP22Metadata::token_name", vec![], None).unwrap()
+	}
+
+	pub(super) fn token_symbol(session: &mut Session<Sandbox>) -> Option<String> {
+		call::<Option<String>>(session, "PSP22Metadata::token_symbol", vec![], None).unwrap()
+	}
+
+	pub(super) fn token_decimals(session: &mut Session<Sandbox>) -> u8 {
+		call::<u8>(session, "PSP22Metadata::token_decimals", vec![], None).unwrap()
+	}
+
+	/// PSP22Mintable test utilities.
+
+	pub(super) fn mint(
+		session: &mut Session<Sandbox>,
+		account: AccountId32,
+		amount: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22Mintable::mint",
+			vec![account.to_string(), amount.to_string()],
+			None,
+		)
+	}
+
+	/// PSP22Burnable test utilities.
+
+	pub(super) fn burn(
+		session: &mut Session<Sandbox>,
+		account: AccountId32,
+		amount: Balance,
+	) -> Result<(), PSP22Error> {
+		call::<()>(
+			session,
+			"PSP22Burnable::burn",
+			vec![account.to_string(), amount.to_string()],
+			None,
+		)
+	}
+
+	/// Call a contract method and decode the returned data.
+	pub(super) fn call<T: Decode>(
+		session: &mut Session<Sandbox>,
+		func_name: &str,
+		input: Vec<String>,
+		endowment: Option<Balance>,
+	) -> Result<T, PSP22Error> {
+		match session.call::<String, ()>(func_name, &input, endowment) {
+			// If the call is reverted, decode the error into `PSP22Error`.
+			Err(SessionError::CallReverted(error)) => {
+				Err(PSP22Error::decode(&mut &error[2..])
+					.unwrap_or_else(|_| panic!("Decoding failed")))
+			},
+			// If the call is successful, decode the last returned value.
+			Ok(_) => Ok(session
+				.record()
+				.last_call_return_decoded::<T>()
+				.unwrap_or_else(|_| panic!("Expected a return value"))
+				.unwrap_or_else(|_| panic!("Decoding failed"))),
+			// Catch-all for unexpected results.
+			_ => panic!("Expected call to revert or return a value"),
+		}
+	}
+
+	/// Convert into `PSP22Error::Custom` error type.
+	pub(super) fn into_psp22_custom<T: Encode>(err: T) -> PSP22Error {
+		let mut padded_vec = err.encode().to_vec();
+		padded_vec.resize(4, 0);
+		// Convert the `Vec<u8>`` to value of `StatusCode`.
+		let array: [u8; 4] = padded_vec.try_into().map_err(|_| "Invalid length").unwrap();
+		let status_code = u32::from_le_bytes(array);
+		PSP22Error::Custom(status_code.to_string())
+	}
+
+	/// This is used to resolve type mismatches between the `AccountId` in the quasi testing
+	/// environment and the contract environment.
+	pub(super) fn account_id_from_slice(s: &[u8; 32]) -> AccountId {
+		AccountId::decode(&mut &s[..]).expect("Should be decoded to AccountId")
+	}
+
+	/// Get the last event from pallet contracts.
+	pub(super) fn last_contract_event(session: &Session<Sandbox>) -> Option<Vec<u8>> {
```

Comment:
If possible, should be moved to pop-drink as generic functions taking account id and error.  Defining this in every example, or contract, adds complexity and developer burden.

Reply:
Done